tool-openssl/req.cc

```diff
@@ -0,0 +1,673 @@
+              type.c_str());
+      // Skip unknown attributes
+      continue;
+    }
+
+    // Skip empty values
+    if (value.empty()) {
+      fprintf(stderr,
+              "Warning: No value specified for attribute \"%s\", Skipped\n",
+              type.c_str());
+      continue;
+    }
+
+    // Add entry to the name
+    if (!X509_NAME_add_entry_by_NID(name.get(), nid, DEFAULT_CHAR_TYPE,
+                                    (unsigned char *)value.c_str(), -1, -1,
+                                    0)) {
+      OPENSSL_PUT_ERROR(X509, ERR_R_X509_LIB);
+      return nullptr;
+    }
+  }
+
+  return name;
+}
+
+typedef struct {
+  const char *field_name;
+  const char *short_desc;
+  const char *default_value;
+  int nid;
+} ReqField;
+
+static const char *prompt_field(const ReqField &field, char *buffer,
+                                size_t buffer_size) {
+  // Prompt with default value if available
+  if (field.default_value && field.default_value[0]) {
+    fprintf(stdout, "%s [%s]: ", field.short_desc, field.default_value);
+  } else {
+    fprintf(stdout, "%s: ", field.short_desc);
+  }
+  fflush(stdout);
+
+  // Get input with fgets
+  if (fgets(buffer, buffer_size, stdin) == NULL) {
+    fprintf(stderr, "Error reading input\n");
+    return NULL;
+  }
+
+  // Remove newline character if present
+  size_t len = OPENSSL_strnlen(buffer, buffer_size);
+  if (len > 0 && buffer[len - 1] == '\n') {
+    buffer[len - 1] = '\0';
+    len--;
+  }
+
+  if (strcmp(buffer, ".") == 0) {
+    // Empty entry requested
+    return "";
+  }
+  if (len == 0 && field.default_value) {
+    return field.default_value;
+  }
+  if (len > 0) {
+    // Use provided input
+    return buffer;
+  }
+
+  // Empty input and no default - use empty string
+  return "";
+}
+
+static bssl::UniquePtr<X509_NAME> prompt_for_subject(
+    X509_REQ *req, bool isCSR, unsigned long chtype = MBSTRING_ASC) {
+  // Default values for subject fields
+  const ReqField subject_fields[] = {
+      {"countryName", "Country Name (2 letter code)", "AU", NID_countryName},
+      {"stateOrProvinceName", "State or Province Name (full name)",
+       "Some-State", NID_stateOrProvinceName},
+      {"localityName", "Locality Name (eg, city)", "", NID_localityName},
+      {"organizationName", "Organization Name (eg, company)",
+       "Internet Widgits Pty Ltd", NID_organizationName},
+      {"organizationalUnitName", "Organizational Unit Name (eg, section)", "",
+       NID_organizationalUnitName},
+      {"commonName", "Common Name (e.g. server FQDN or YOUR name)", "",
+       NID_commonName},
+      {"emailAddress", "Email Address", "", NID_pkcs9_emailAddress}};
+
+  // Extra attributes for CSR
+  const ReqField extra_attributes[] = {
+      {"challengePassword", "A challenge password", "",
+       NID_pkcs9_challengePassword},
+      {"unstructuredName", "An optional company name", "",
+       NID_pkcs9_unstructuredName}};
```

Comment:
NP: The function would flow better if these constants were moved to be right before the function.